### PR TITLE
Perf/164861 ess performance readmetxt cache

### PR DESCRIPTION
**Pull request by Pradnya-Bhosale**
This PR contains, Implementation and respective log-events, unit test changes  for invalidating and caching of readme.txt file by extending existing webhook .

Thanks,

### CODE REVIEW

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentDataService.cs [hunk 1]
```diff
@@ -260,16 +260,39 @@ private async Task CreateAncillaryFiles(string batchId, string exchangeSetPath,
         }
 
         public async Task<bool> DownloadReadMeFile(string batchId, string exchangeSetRootPath, string correlationId)
```
Comment (Sums13943) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentDataService.cs:262]: This method assumes success as long as no exceptions are thrown. Consider adding error handling to manage cases where the methods fail unexpectedly. 
Unlike the other method _DownloadReadMeFileFromFss_, which involves an API call and may return a success code of false.
Also add unit test case to cover the same.

**Review by Sums13943 — COMMENTED** [review]
few minor comments

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/IFileShareServiceCache.cs [hunk 2]
```diff
@@ -13,5 +13,6 @@ public interface IFileShareServiceCache
         Task<List<Products>> GetNonCachedProductDataForFss(List<Products> products, SearchBatchResponse internalSearchBatchResponse, string exchangeSetRootPath, SalesCatalogueServiceResponseQueueMessage queueMessage, CancellationTokenSource cancellationTokenSource, CancellationToken cancellationToken, string businessUnit);
         Task CopyFileToBlob(Stream stream, string fileName, string batchId);
         Task InsertOrMergeFssCacheDetail(FssSearchResponseCache fssSearchResponseCache);
+        Task<Stream> DownloadFileFromCache(string fileName, string containerName);
```
[the remark this answers is not in the document]
Comment (Rashi14748) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/IFileShareServiceCache.cs:16]: This is generic method to download blob, not specific to ReadMe cache.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/Helper/ClearCacheHelper.cs [hunk 3]
```diff
@@ -31,6 +34,24 @@ private async Task<object> ExecuteTableOperation(TableOperation tableOperation,
             return tableResult.Result;
         }
 
+        public async Task<bool> GetProductIdentifier(string essJwtToken, string essBaseAddress, string readmeContainer, string connectionString)
```
[the remark this answers is not in the document]
Comment (Sheetal15692) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/Helper/ClearCacheHelper.cs]: Thanks for Reviewing..
Done

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/Helper/FileContentHelper.cs [hunk 4]
```diff
@@ -486,5 +489,25 @@ public static void CheckSerialAioFileContentForAioUpdate(string inputFile)
             Assert.AreEqual(dateAndCdType, $"{currentDate}UPDATE", $"Incorrect date is returned '{currentDate}UPDATE', instead of the expected {dateAndCdType}.");
             Assert.IsTrue(formatVersionAndExchangeSetNumber.StartsWith("02.00"), $"Expected format version {formatVersionAndExchangeSetNumber}");
         }
+
+        public static async Task<bool> WaitForContainerAsync(BlobServiceClient blobServiceClient, string containerName, int maxAttempts, int delayMilliseconds)
```
[the remark this answers is not in the document]
Comment (Sheetal15692) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/Helper/FileContentHelper.cs]: Thanks For Reviewing..
Done

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ClearCacheVerification.cs [hunk 5]
```diff
@@ -142,10 +142,34 @@ public async Task WhenICallExchangeSetApiWithProductIdentifiersForExchangeSetSta
...
+            await ClearCacheHelper.GetProductIdentifier(EssJwtToken, Config.EssBaseAddress, readmeContainer, Config.ClearCacheConfig.CacheStorageConnectionString);
+
+            // newfile publish hit
+            var essCacheJson = ClearCacheHelper.GetDataForPayload(Config.ClearCacheConfig.Source, Config.ClearCacheConfig.Id);
+            essCacheJson["Data"] = JObject.FromObject(ClearCacheHelper.GetCacheRequestDataForReadMeFile(Config.BESSConfig.S63BusinessUnit));
+            var apiClearCacheResponse = await ExchangeSetApiClient.PostNewFilesPublishedAsync(essCacheJson, accessToken: EssJwtToken);
+            Assert.AreEqual(200, (int)apiClearCacheResponse.StatusCode, $"Incorrect status code is returned for clear cache endpoint {apiClearCacheResponse.StatusCode}, instead of the expected status 200.");
+
+            // Verify the no Cache available for readme
+            bool containerExists = await FileContentHelper.WaitForContainerAsync(BlobServiceClient, readmeContainer,3,7000);
```
[the remark this answers is not in the document]
Comment (Sheetal15692) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ClearCacheVerification.cs]: Thanks For Reviewing..
Done

**Review by nevillejrbrown — APPROVED** [review]

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareService.cs [hunk 6]
```diff
@@ -574,6 +576,23 @@ public async Task<bool> DownloadReadMeFile(string readMeFilePath, string batchId
             }
         }
 
+        public async Task<bool> DownloadReadMeFileFromCache(string batchId, string exchangeSetRootPath, string correlationId)
+        {
+            bool isReadMeFileDownloade = false;
```
[the remark this answers is not in the document]
Comment (Pradnya-Bhosale) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareService.cs]: done

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentDataService.cs (hunk 1 above)
Comment (Pradnya-Bhosale) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentDataService.cs:262]: added exception handling logic

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareServiceCache.cs [hunk 7]
```diff
@@ -231,5 +231,19 @@ public async Task InsertOrMergeFssCacheDetail(FssSearchResponseCache fssSearchRe
             var storageConnectionString = azureStorageService.GetStorageAccountConnectionString(fssCacheConfiguration.Value.CacheStorageAccountName, fssCacheConfiguration.Value.CacheStorageAccountKey);
             await azureTableStorageClient.InsertOrMergeIntoTableStorageAsync(fssSearchResponseCache, fssCacheConfiguration.Value.FssSearchCacheTableName, storageConnectionString);
         }
+
+        public async Task<Stream> DownloadFileFromCache(string fileName, string containerName)
+        {
+            MemoryStream memoryStream = new MemoryStream();
+            var storageConnectionString = azureStorageService.GetStorageAccountConnectionString(fssCacheConfiguration.Value.CacheStorageAccountName, fssCacheConfiguration.Value.CacheStorageAccountKey);
+
+            CloudBlockBlob cloudBlockBlob = azureBlobStorageClient.GetExistingCloudBlockBlob(fileName, storageConnectionString, containerName);
```
[the remark this answers is not in the document]
Comment (Pradnya-Bhosale) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareServiceCache.cs]: as suggested in one of the review comment removed the above method, so marking this comment as Resolved.
Thanks!

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/AzureBlobStorageClient.cs [hunk 8]
```diff
@@ -53,5 +53,14 @@ public async Task DeleteCacheContainer(string storageAccountConnectionString, st
...
         }
+
+        public CloudBlockBlob GetExistingCloudBlockBlob(string fileName, string storageAccountConnectionString, string containerName)
+        {
+            CloudStorageAccount cloudStorageAccount = CloudStorageAccount.Parse(storageAccountConnectionString);
+            CloudBlobClient cloudBlobClient = cloudStorageAccount.CreateCloudBlobClient();
+            CloudBlobContainer cloudBlobContainer = cloudBlobClient.GetContainerReference(containerName);
+            CloudBlockBlob cloudBlockBlob = cloudBlobContainer.GetBlockBlobReference(fileName);
+            return cloudBlockBlob;
+        }
```
[the remark this answers is not in the document]
Comment (Pradnya-Bhosale) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/AzureBlobStorageClient.cs]: modified the existing method with one parameter.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileSystemHelper.cs [hunk 9]
```diff
@@ -94,6 +94,7 @@ public bool DownloadReadmeFile(string filePath, Stream stream, string lineToWrit
         {
             if (stream != null)
             {
+                stream.Position= 0;
```
[the remark this answers is not in the document]
Comment (Pradnya-Bhosale) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileSystemHelper.cs]: done

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareService.cs [hunk 10]
```diff
@@ -574,6 +576,23 @@ public async Task<bool> DownloadReadMeFile(string readMeFilePath, string batchId
...
+            string fileName = fileShareServiceConfig.Value.ReadMeFileName;
+            string filePath = Path.Combine(exchangeSetRootPath, fileName);
+            fileSystemHelper.CheckAndCreateFolder(exchangeSetRootPath);
+            string lineToWrite = string.Concat("File date: ", DateTime.UtcNow.ToString("yyyy-MM-dd HH:mm:ssZ", CultureInfo.InvariantCulture));            
+            using (Stream readmeStream = await fileShareServiceCache.DownloadFileFromCache(fileName, ReadMeContainerName))
+            {
+                if (readmeStream.Length > 0)
```
[the remark this answers is not in the document]
Comment (Pradnya-Bhosale) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareService.cs]: added null check

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareServiceCache.cs [hunk 11]
```diff
@@ -231,5 +231,19 @@ public async Task InsertOrMergeFssCacheDetail(FssSearchResponseCache fssSearchRe
             var storageConnectionString = azureStorageService.GetStorageAccountConnectionString(fssCacheConfiguration.Value.CacheStorageAccountName, fssCacheConfiguration.Value.CacheStorageAccountKey);
             await azureTableStorageClient.InsertOrMergeIntoTableStorageAsync(fssSearchResponseCache, fssCacheConfiguration.Value.FssSearchCacheTableName, storageConnectionString);
         }
+
+        public async Task<Stream> DownloadFileFromCache(string fileName, string containerName)
+        {
+            MemoryStream memoryStream = new MemoryStream();
```
[the remark this answers is not in the document]
Comment (Pradnya-Bhosale) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareServiceCache.cs]: memory disposal added in subsequent method i.e DownloadReadMeFileFromCache

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentDataService.cs [hunk 12]
```diff
@@ -260,16 +260,39 @@ private async Task CreateAncillaryFiles(string batchId, string exchangeSetPath,
...
+        {
+            bool isDownloadReadMeFileSuccess = false;
+            DateTime createReadMeFileTaskStartedAt = DateTime.UtcNow;
+
+            logger.LogInformation(EventIds.SearchDownloadReadmeCacheEventStart.ToEventId(), "Cache Search and Download readme.txt file started for BatchId:{BatchId} and _X-Correlation-ID:{CorrelationId}", batchId, correlationId);
+
+            isDownloadReadMeFileSuccess = await fulfilmentFileShareService.DownloadReadMeFileFromCache(batchId, exchangeSetRootPath, correlationId);
```
[the remark this answers is not in the document]
Comment (Pradnya-Bhosale) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentDataService.cs]: after adding code for exception handling isDownloadReadMeFileSuccess variable needs to define explicitly for return value

**Review by DanOwenUKHO — CHANGES_REQUESTED** [review]
Two minor changes

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentDataService.cs [hunk 13]
```diff
@@ -260,16 +260,48 @@ private async Task CreateAncillaryFiles(string batchId, string exchangeSetPath,
...
+                    monitorHelper.MonitorRequest("Download ReadMe File Task", createReadMeFileTaskStartedAt, createReadMeFileTaskCompletedAt, correlationId, null, null, null, batchId);
+                }
+                else
+                {
+                    logger.LogInformation(EventIds.ReadMeTextFileNotFound.ToEventId(), "Cache Search and Download readme.txt file not found for BatchId:{BatchId} and _X-Correlation-ID:{CorrelationId}", batchId, correlationId);
+                    isDownloadReadMeFileSuccess = await DownloadReadMeFileFromFss(batchId, exchangeSetRootPath, correlationId);
+                }
+
+                if(!isDownloadReadMeFileSuccess)
+                    logger.LogError(EventIds.ErrorInDownloadReadMeFile.ToEventId(), "Error while downloading readme.txt file for BatchId:{BatchId} and _X-Correlation-ID:{CorrelationId}", batchId, correlationId);
```
[the remark this answers is not in the document]
Comment (Pradnya-Bhosale) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentDataService.cs:286]: added braces

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/appsettings.json [hunk 14]
```diff
@@ -47,7 +47,9 @@
 
   "CacheConfiguration": {
     "CacheStorageConnectionString": "",
-    "FssSearchCacheTableName": "FssSearchResponseCache"
+    "FssSearchCacheTableName": "FssSearchResponseCache",
+    "Source": "AcceptanceTest",
+    "Id": "25d6c6c1-418b-40f9-bb76-f6dfc0f133bc"
```
[the remark this answers is not in the document]
Comment (Sheetal15692) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/appsettings.json:52]: Thanks for reviewing..
Its not a blob id Its just id comes with newFilesPublished event payload and we used that same payload and just added values for mock.

**Review by harshal11869 — APPROVED** [review]
Looks Good.

**Review by DanOwenUKHO — APPROVED** [review]

**Review by amogh14893 — APPROVED** [review]
LGTM. apporved

**Review by Sums13943 — APPROVED** [review]

**Pull request merged**